src/Main.elm

```diff
@@ -1,10 +1,209 @@
+main : Program () Model Msg
 main =
-    Html.text "There will be app soon!"
+    Browser.sandbox
+        { init = init
+        , view = view
+        , update = update
+        }
+
+
+type Msg
+    = FolderClick Int
+    | GoUpClick
+
+
+type alias Model =
+    Result String GoodModel
+
+
+{-| The integer IDs are not the same as the ones provided in the data.
+It seems likely that the provided IDs are row IDS from the database,
+and that folders and audiences are kept in different tables, so
+the IDs may well conflict. A globally unique ID is needed here.
+
+This model design is inspired by the book Data-Oriented Design by
+Richard Fabian, see <https://www.dataorienteddesign.com/dodbook> .
+It's an idea from game dev that says that using tabular data
+structures gives better performance. I am experimenting with using
+it in high-level languages, like Elm and Go. I am finding so far
+that it gives a really elegant and minimal internal state.
+
+-}
+type alias GoodModel =
+    -- The keys are child IDs and the values are parent IDs. Root
+    -- items are not included. If a map lookup fails then the item
+    -- is root, so there are no internal errors caused by failing map
+    -- lookups. The reason for using a dictionary rather than a list
+    -- is to enforce that a child can have only one parent.
+    { parents : Dict.Dict Int Int
+
+    -- As well as providing the AudienceType for audiences, this is
+    -- used to detect whether something is a folder or an audience.
+    , audiences : Dict.Dict Int A.AudienceType
+
+    -- The IDs and names for all the folders and audiences. This should
+    -- only be mapped over, not used for lookup. Lookup is bad in this
+    -- case because failure will have to be handled, and it shouldn't
+    -- ever be possible to have an ID without a name. The reason for
+    -- using a dictionary instead of a list is to enforce that an
+    -- item can only have one name.
+    , all : Dict.Dict Int String
+    , parent : Parent
+    }
+
+
+type Parent
+    = Root
+    | Parent Int
+
+
+view : Model -> Html.Html Msg
+view model =
+    case model of
+        Err err ->
+            Html.text <| "internal error: " ++ err
+
+        Ok good ->
+            viewGood good
+
+
+viewGood : GoodModel -> Html.Html Msg
+viewGood model =
+    Html.div
+        [ Hat.id "container" ]
+    <|
+        List.concat
+            [ goUpView model
+            , foldersView model
+            , audiencesView model
+            ]
+
+
+goUpView : GoodModel -> List (Html.Html Msg)
+goUpView model =
+    if model.parent == Root then
+        []
+
+    else
+        [ Html.button
+            [ Hev.onClick GoUpClick ]
+            [ Html.text "Go up" ]
+        ]
+
+
+foldersView : GoodModel -> List (Html.Html Msg)
+foldersView model =
+    List.map oneFolderView <|
+        Dict.toList <|
+            Dict.filter (isFolder model.audiences) <|
+                Dict.filter (isChildOf model.parents model.parent) model.all
```

Comment:
Yours looks neater visually, but I find it harder to understand. I think it's all the years of using languages that have the arguments in parenthesis has made me feel that arguments must go on the right. Perhaps I should try to get used to it this way.

Reply:
I definitely didn't mean it as "you SHOULD do it this way"! :) it's just that I find it nicer and wanted to share that possibility in case you didn't know it.

Reply:
The more I look at it the more I like it.